### ISSUE TITLE
Allow exporting CSS for SSR when using styled-jsx for Components Library

### ISSUE THREAD

**Issue by essential-randomness** [zeit/styled-jsx#646]
#### Do you want to request a *feature* or report a *bug*?

Feature (unless it's somehow possible already and I just couldn't figure it out).

#### What is the current behavior?

No way to export static CSS for SSR when using styled-jsx in a standalone library.

This is the only other person seemingly attempting this I could find, and it seems like they gave up: https://stackoverflow.com/questions/55050039/intense-fouc-when-using-next-js-8-and-styled-jsx

#### If the current behavior is a bug, please provide the steps to reproduce and possibly a minimal demo or testcase in the form of a Next.js app, CodeSandbox URL or similar

What I'm doing in case there's something I'm not seeing:

1. Created a standalone UI components library. The components are styled using styled-jsx, then built into a library.
2. Created a NextJS app that imports the component library. 
3. The styling works, but I get FOUC because the CSS in the component library is not applied server side (and takes a while to load)

Stuff I attempted:

** ATTEMPT 1 **

Added this code in the library in a SSRStyles file exported with it:
```
export default () => {
  const app = ReactDOM.renderToString(
    <div>
      <Layout />
      <Button />
    </div>
  );
  const styles = flush().map(
    (style) =>
      style.props.dangerouslySetInnerHTML.__html
  );
  return styles;
};
```

This worked somewhat but a) is ugly, b) I need to do this for every component I export, c) it somehow doesn't even have all the styles (but it's way better than without).

** ATTEMPT 2**

Added this code in the NextJS app, in _document.tsx:

```
  static async getInitialProps(ctx: any) {
    const initialProps = await Document.getInitialProps(ctx);

    console.log("*\n*\n*\n*\n*\n*\n*\n");
    const app = ReactDOM.renderToString(<HomePage />);
    console.log(app);
    const styles = flushToHTML();
    console.log(styles);
    return { ...initialProps };
  }
```

Styles is empty. Using ctx.AppTree() similarly gives empty styles. Importing components from my library and server-side rendering those also gives empty styles.

I've tried a variety of combinations for this and couldn't get it to work.

#### What is the expected behavior?

There is an easy way to extract all the styles for SSR. Without this I'm unsure how styled-jsx can be used to create component libraries.

#### Environment (include versions)

* OS: MacOSX
* Browser: All
* styled-jsx (version): NextJS App: "next": "^9.3.6", Components Library:  "styled-jsx": "^3.2.5"
* If you want to see how I'm packaging the library, this is another project I have in which I'm doing the same (the other is private): https://github.com/essential-randomness/boba-editor 

#### Did this work in previous versions?

No idea.

**Comment by giuseppeg** [comment]
Static extraction right now is not possible because styled-jsx supports dynamic styles and styles composition. SSR works like explained in the README, if you are building an app that is not hydrated on the client and you are willing to manage `link` tags insertion then you can render the styles on the server on a per page level.

**Comment by essential-randomness** [comment]
As mentioned in the issue, I have tried to do SSR as explained in the README, but I don't get anything returned from methods like "flush()". I suspect that this is because I separated the components into their own library/repo and I import them in my project after compilation. Is this working as expected, or is there something I'm doing wrong?

**Comment by essential-randomness** [comment]
To expand, what I'd like to happen is:

1. I create my components library which is styled through styled-jsx
2. I import my component library in another frontend project which supports SSR (in this case NextJS)
3. I do some SSR magic and I'm able to avoid FOUC when using my component library.


Even if static exports are not the solution to this issue, I'd like to understand whether this is something that's supported at all.

**Comment by giuseppeg** [comment]
yep https://medium.com/@tomaszmularczyk89/guide-to-building-a-react-components-library-with-rollup-and-styled-jsx-694ec66bd2

**Comment by giuseppeg** [comment]
the consumer app needs to install styled-jsx and probably you need to configure https://github.com/zeit/styled-jsx#stylemodule when building the components library

**Comment by essential-randomness** [comment]
Thank you, the bit about styleModule seems promising, since I haven't tried anything related yet (since the consumer app is NextJS styled-jsx is already installed).

Is there any further documentation on what the options are/what styleModule is? I see this is a very new option and I'm uncertain where to even begin fiddling with it.

**Comment by essential-randomness** [comment]
Reporting back after another 4-5 hours trying to solve this problem. I got somewhere. It's not pretty, but I got somewhere.

I have abandoned pursuing the styleModule route because there's no documentation at all that I could find. It seems promising, but even reading the source code didn't leave me with any clue on where to start investigating.

The solution came from a [2018 NextJS issue](https://github.com/zeit/next.js/issues/3623) where someone mentioned exporting the `flush` method of the component library to use in the `_document.js` of the consumer app. This got things to work on my localhost, both in dev and prod mode.

For some reason, however, I was still having trouble when deploying to cloud (I'm using gCloud for this). The server complained about being unable to find my components library even though it was working perfectly fine on the client (note: this might be due to using `yarn link`). To solve this, I installed the [next-transpile-modules](https://github.com/martpie/next-transpile-modules) Babel plugin. After getting the dreaded ["multiple react instances error"](https://github.com/facebook/react/issues/13991) and going down another configurations rabbit-hole I got to the following `next.config.js`  for the consumer app:

```
const withTM = require("next-transpile-modules")(["MY_COMPONENTS_LIBRARY"]);
const path = require("path");

module.exports = withTM({
  webpack: (config, options) => {
    config.resolve.alias["react"] = path.resolve(
      __dirname,
      ".",
      "node_modules",
      "react"
    );
    config.resolve.alias["react-dom"] = path.resolve(
      __dirname,
      ".",
      "node_modules",
      "react-dom"
    );
    if (options.isServer) {
      config.externals = ["react", "react-dom", ...config.externals];
    }
    return config;
  },
});

```

...and that was it. 

All of this feels extremely brittle. My gut feeling is the root cause is probably related to #64, which is where my research always led back to. Hopefully once that's fixed there is a more intuitive way to get this to work.

Hope all this sleuthing helps someone!
